### PR TITLE
improved flavor docs

### PR DESCRIPTION
**Pull request by htahir1**
## Describe changes
They bugged me.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: docs/book/advanced-guide/stacks/custom-flavors.md [hunk 1]
```diff
@@ -220,12 +220,31 @@ class MyArtifactStoreFlavor(BaseArtifactStoreFlavor):
...
 
-Once your implementation is complete, you can register it through the CLI:
+Once your implementation is complete, you can register it through the CLI.
+Please ensure you **point to the flavor class via dot notation**:
 
 ```shell
 zenml artifact-store flavor register <path.to.MyArtifactStoreFlavor>
 ```
 
+For example, your flavor class `MyArtifactStoreFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/advanced-guide/stacks/custom-flavors.md:230]:
```suggestion
For example, if your flavor class `MyArtifactStoreFlavor` is defined in `flavors/my_flavor.py`,
```

File: docs/book/advanced-guide/stacks/custom-flavors.md [hunk 2]
```diff
@@ -220,12 +220,31 @@ class MyArtifactStoreFlavor(BaseArtifactStoreFlavor):
...
+you'd register it by doing:
+
+```shell
+zenml artifact-store flavor register flavors.my_flavor.MyArtifactStoreFlavor
+```
+
+{% hint style="warning" %}
+ZenML resolves the flavor class by taking the path where you initialized zenml
```
Comment (strickvl) [inline docs/book/advanced-guide/stacks/custom-flavors.md:238]:
```suggestion
ZenML resolves the flavor class by taking the path where you initialized ZenML
```

File: docs/book/advanced-guide/stacks/custom-flavors.md [hunk 3]
```diff
@@ -220,12 +220,31 @@ class MyArtifactStoreFlavor(BaseArtifactStoreFlavor):
...
+(via `zenml init`) as the starting point of resolution. Therefore, please ensure
+you follow [the best practice](../../guidelines/best-practices.md) of initializing
+zenml at the root of your respository.
```
Comment (strickvl) [inline docs/book/advanced-guide/stacks/custom-flavors.md:241]:
```suggestion
ZenML at the root of your respository.
```

File: docs/book/advanced-guide/stacks/custom-flavors.md [hunk 4]
```diff
@@ -220,12 +220,31 @@ class MyArtifactStoreFlavor(BaseArtifactStoreFlavor):
...
+
+If ZenML does not find an initialized ZenML repository in any parent directory, it
+will default to the current working directory, but usually its better to not have to
+rely on this mechanism, and initialize zenml at the root.
```
Comment (strickvl) [inline docs/book/advanced-guide/stacks/custom-flavors.md:245]:
```suggestion
rely on this mechanism, and initialize ZenML at the root.
```

File: docs/book/component-gallery/alerters/custom.md [hunk 5]
```diff
@@ -111,10 +111,36 @@ class MyAlerterFlavor(BaseAlerterFlavor):
...
-through the CLI as:
+through the CLI.
+Please ensure you **point to the flavor class via dot notation**: 
 
 ```shell
-zenml alerter flavor register <THE-SOURCE-PATH-OF-YOUR-ALERTER-FLAVOR>
+zenml alerter flavor register <path.to.MyAlerterFlavor>
+```
+
+For example, your flavor class `MyAlerterFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/component-gallery/alerters/custom.md:121]:
```suggestion
For example, if your flavor class `MyAlerterFlavor` is defined in `flavors/my_flavor.py`,
```

File: docs/book/component-gallery/artifact-stores/custom.md [hunk 6]
```diff
@@ -168,11 +168,36 @@ and fill in the `SUPPORTED_SCHEMES` based on your file system.
...
-as:
+Once you are done with the implementation, you can register it through the CLI.
+Please ensure you **point to the flavor class via dot notation**: 
 
 ```shell
-zenml artifact-store flavor register <THE-SOURCE-PATH-OF-YOUR-ARTIFACT-STORE-FLAVOR>
+zenml artifact-store flavor register <path.to.MyArtifactStoreFlavor>
+```
+
+For example, your flavor class `MyArtifactStoreFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/component-gallery/artifact-stores/custom.md:178]:
```suggestion
For example, if your flavor class `MyArtifactStoreFlavor` is defined in `flavors/my_flavor.py`,
```

File: docs/book/component-gallery/container-registries/custom.md [hunk 7]
```diff
@@ -89,11 +89,35 @@ from the `BaseContainerRegistryConfig` class.
...
-as:
+Once you are done with the implementation, you can register it through the CLI.
+Please ensure you **point to the flavor class via dot notation**: 
 
 ```shell
-zenml container-registry flavor register <THE-SOURCE-PATH-OF-YOUR-CONTAINER-REGISTRY-FLAVOR>
+zenml container-registry flavor register <path.to.MyContainerRegistryFlavor>
+```
+
+For example, your flavor class `MyContainerRegistryFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/component-gallery/container-registries/custom.md:99]:
```suggestion
For example, if your flavor class `MyContainerRegistryFlavor` is defined in `flavors/my_flavor.py`,
```

File: docs/book/component-gallery/data-validators/custom.md [hunk 8]
```diff
@@ -37,11 +37,35 @@ from the `BaseDataValidatorConfig` class.
...
-as:
+Once you are done with the implementation, you can register it through the CLI.
+Please ensure you **point to the flavor class via dot notation**: 
 
 ```shell
-zenml data-validator flavor register <THE-SOURCE-PATH-OF-YOUR-DATA-VALIDATOR-FLAVOR>
+zenml data-validator flavor register <path.to.MyDataValidatorFlavor>
+```
+
+For example, your flavor class `MyDataValidatorFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/component-gallery/data-validators/custom.md:47]:
```suggestion
For example, if your flavor class `MyDataValidatorFlavor` is defined in `flavors/my_flavor.py`,
```

File: docs/book/component-gallery/experiment-trackers/custom.md [hunk 9]
```diff
@@ -27,11 +27,36 @@ implement the abstract methods.
...
-as:
+Once you are done with the implementation, you can register it through the CLI.
+Please ensure you **point to the flavor class via dot notation**: 
 
 ```shell
-zenml experiment-tracker flavor register <THE-SOURCE-PATH-OF-YOUR-EXPERIMENT-TRACKER-FLAVOR>
+zenml experiment-tracker flavor register <path.to.MyExperimentTrackerFlavor>
+```
+
+For example, your flavor class `MyExperimentTrackerFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/component-gallery/experiment-trackers/custom.md:37]:
```suggestion
For example, if your flavor class `MyExperimentTrackerFlavor` is defined in `flavors/my_flavor.py`,
```

File: docs/book/component-gallery/image-builders/custom.md [hunk 10]
```diff
@@ -82,11 +82,35 @@ from the `BaseImageBuilderConfig` class add your configuration parameters.
...
-as:
+Once you are done with the implementation, you can register it through the CLI.
+Please ensure you **point to the flavor class via dot notation**: 
 
 ```shell
-zenml image-builder flavor register <IMAGE-BUILDER-FLAVOR-SOURCE-PATH>
+zenml image-builder flavor register <path.to.MyImageBuilderFlavor>
+```
+
+For example, your flavor class `MyImageBuilderFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/component-gallery/image-builders/custom.md:92]:
```suggestion
For example, if your flavor class `MyImageBuilderFlavor` is defined in `flavors/my_flavor.py`,
```

File: docs/book/component-gallery/model-deployers/custom.md [hunk 11]
```diff
@@ -178,11 +178,35 @@ from the `BaseModelDeployerConfig` class add your configuration parameters.
...
-as:
+Once you are done with the implementation, you can register it through the CLI.
+Please ensure you **point to the flavor class via dot notation**: 
 
 ```shell
-zenml model-deployer flavor register <THE-SOURCE-PATH-OF-YOUR-MODEL_DEPLOYER-FLAVOR>
+zenml model-deployer flavor register <path.to.MyModelDeployerFlavor>
+```
+
+For example, your flavor class `MyModelDeployerFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/component-gallery/model-deployers/custom.md:188]:
```suggestion
For example, if your flavor class `MyModelDeployerFlavor` is defined in `flavors/my_flavor.py`,
```

File: docs/book/component-gallery/orchestrators/custom.md [hunk 12]
```diff
@@ -78,11 +78,35 @@ from the `BaseOrchestratorConfig` class add your configuration parameters.
...
-as:
+Once you are done with the implementation, you can register it through the CLI.
+Please ensure you **point to the flavor class via dot notation**: 
 
 ```shell
-zenml orchestrator flavor register <THE-SOURCE-PATH-OF-YOUR-ORCHESTRATOR-FLAVOR>
+zenml orchestrator flavor register <path.to.MyOrchestratorFlavor>
+```
+
+For example, your flavor class `MyOrchestratorFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/component-gallery/orchestrators/custom.md:88]:
```suggestion
For example, if your flavor class `MyOrchestratorFlavor` is defined in `flavors/my_flavor.py`,
```

File: docs/book/component-gallery/secrets-managers/custom.md [hunk 13]
```diff
@@ -97,11 +97,36 @@ from the `BaseSecretsManagerConfig` class add your configuration parameters.
...
-as:
+
+Once you are done with the implementation, you can register it through the CLI.
+Please ensure you **point to the flavor class via dot notation**: 
+
+```shell
+zenml secrets-manager flavor register <path.to.MySecretsManagerFlavor>
+```
+
+For example, your flavor class `MySecretsManagerFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/component-gallery/secrets-managers/custom.md:108]:
```suggestion
For example, if your flavor class `MySecretsManagerFlavor` is defined in `flavors/my_flavor.py`,
```

File: docs/book/component-gallery/step-operators/custom.md [hunk 14]
```diff
@@ -103,11 +103,36 @@ from the `BaseOrchestratorConfig` class add your configuration parameters.
...
-as:
+
+Once you are done with the implementation, you can register it through the CLI.
+Please ensure you **point to the flavor class via dot notation**: 
+
+```shell
+zenml step-operator flavor register <path.to.MyStepOperatorFlavor>
+```
+
+For example, your flavor class `MyStepOperatorFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/component-gallery/step-operators/custom.md:114]:
```suggestion
For example, if your flavor class `MyStepOperatorFlavor` is defined in `flavors/my_flavor.py`,
```

File: docs/book/misc/integrating.md [hunk 15]
```diff
@@ -45,8 +45,32 @@ and use/test them as a regular custom flavor. E.g., when
 [developing an orchestrator](../component-gallery/orchestrators/custom.md) 
 you can use:
 
+```shell
+zenml orchestrator flavor register <path.to.MyOrchestratorFlavor>
+```
+
+For example, your flavor class `MyOrchestratorFlavor` is defined in `flavors/my_flavor.py`,
```
Comment (strickvl) [inline docs/book/misc/integrating.md:52]:
```suggestion
For example, if your flavor class `MyOrchestratorFlavor` is defined in `flavors/my_flavor.py`,
```

**Review by strickvl — CHANGES_REQUESTED** [review]
Wonderful. Had it on my list to make some changes based on yesterday's conversation this morning too!

File: docs/book/advanced-guide/stacks/custom-flavors.md [hunk 16]
```diff
@@ -220,12 +220,31 @@ class MyArtifactStoreFlavor(BaseArtifactStoreFlavor):
...
+```
+
+{% hint style="warning" %}
+ZenML resolves the flavor class by taking the path where you initialized zenml
+(via `zenml init`) as the starting point of resolution. Therefore, please ensure
+you follow [the best practice](../../guidelines/best-practices.md) of initializing
+zenml at the root of your respository.
+
+If ZenML does not find an initialized ZenML repository in any parent directory, it
+will default to the current working directory, but usually its better to not have to
```
Comment (schustmi) [inline docs/book/advanced-guide/stacks/custom-flavors.md:244]: I think this is incorrect right? It doesn't use the current working directory, but the parent directory of the python main module (If running `python run.py`, these will be the same as the working dir. You could also do `python some_dir/run.py`, in which case they won't match)

**Review by schustmi — CHANGES_REQUESTED** [review]

File: docs/book/advanced-guide/stacks/custom-flavors.md (hunk 16 above)
Comment (schustmi) [reply docs/book/advanced-guide/stacks/custom-flavors.md:244]: Actually in the case of the flavor command, even that is incorrect as the main module will be the ZenML CLI script.

Comment (schustmi) [reply docs/book/advanced-guide/stacks/custom-flavors.md:244]: I'll have a call with Stefan later where will discuss the same issue for `zenml pipeline build/run`, I'll report back what we're going to do in this case.

Comment (schustmi) [reply docs/book/advanced-guide/stacks/custom-flavors.md:244]: What we decided: Your doc update here regarding the working directory is the way we want this to work, but not the way it's implemented right now. So we'd also have to fix the code to actually make it work that way

Comment (schustmi) [reply docs/book/advanced-guide/stacks/custom-flavors.md:244]: Continuing my conversation here: I'll be fixing this in a generic way on my branch, no need to do anything here

Comment (htahir1) [reply docs/book/advanced-guide/stacks/custom-flavors.md:244]: @schustmi Should I then go ahead and merge this when its ready or will your branch not be ready before next release?

Comment (schustmi) [reply docs/book/advanced-guide/stacks/custom-flavors.md:244]: Go for it

**Review by strickvl — APPROVED** [review]

**Review by fa9r — APPROVED** [review]

**Review by bcdurak — APPROVED** [review]

**Pull request merged**